File: my-folder/0153-find-minimum-in-rotated-sorted-array/solution.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        if(nums.size() == 1) {
            return nums[0];
        }
        int right = nums.size() - 1;
        int left = 0;
        

        if(nums[right] > nums[0]) {
            return nums[0];
        }

        while(right >= left) {
            int pointer = left + (right - left) / 2;

            if(nums[pointer] > nums[pointer + 1]) {
                return nums[pointer + 1];
            }

            if(nums[pointer - 1] > nums[pointer]) {
                return nums[pointer];
            }

            if(nums[pointer] > nums[0]) {
                left = pointer + 1;
            } else {
                right = pointer - 1;
            }
        }

        return -1;
    }
};

File: my-folder/0153-find-minimum-in-rotated-sorted-array/solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMin = function(nums) {
    let l = 0, r = nums.length - 1;
    let result = nums[0];

    while (l <= r) {
        if(nums[l] < nums[r]) {
            result = Math.min(result, nums[l]);
            break;
        }

        let mid = l + Math.floor((r - l) / 2);
        result = Math.min(result, nums[mid]);

        if(nums[mid] >= nums[l]) {
            l = mid + 1;
        } else {
            r = mid - 1;
        }
    }

    return result;
};
